bgpv1: expose bgp peering state via cilium agent cli

harsimran-pabla opened the pull request:
Introduces new CLI `cilium bgp peers`, which fetches BGP peering state from Cilium virtual router.

Sample output
```
root@clab-bgp-cplane-demo-worker:/home/cilium# cilium bgp peers
Local AS   Peer AS   Peer Address   Session       Uptime   Family         Received   Advertised
65010      65010     10.0.0.1       established   35m30s   ipv4_unicast   11         1
                                                           ipv6_unicast   0          0
root@clab-bgp-cplane-demo-worker:/home/cilium#
root@clab-bgp-cplane-demo-worker:/home/cilium# cilium bgp peers -o json
[
  {
    "families": [
      {
        "accepted": 11,
        "advertised": 1,
        "afi": "ipv4",
        "received": 11,
        "safi": "unicast"
      },
      {
        "afi": "ipv6",
        "safi": "unicast"
      }
    ],
    "local-asn": 65010,
    "peer-address": "10.0.0.1",
    "peer-asn": 65010,
    "session-state": "established",
    "uptime": 2133000000000
  }
]
```

```release-note
Provides operational state of BGP peers via CLI 'cilium bgp peers'
```
Signed-off-by: harsimran pabla <hpabla@isovalent.com>

dylandreimerink commented on cilium/cmd/bgp_peer_get.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+func printSummary(peers []*models.BgpPeer) {
+	w := tabwriter.NewWriter(os.Stdout, 5, 0, 3, ' ', 0)
```
Can you break out these magic numbers into consts (can be just above). I have no clue what these numbers do without looking up the docs.

Also, how do other CLI commands handle table generation? Is there a more generic solution or do we do things like this everywhere?

harsimran-pabla replied:
I'll add some constants for it. I took reference from https://github.com/cilium/cilium/blob/master/cilium/cmd/bpf_auth_list.go#L78 

Looking at other commands, it seems like similar approach is used everywhere.

YutaroHayakawa commented on pkg/bgpv1/gobgp/state.go:
```diff
@@ -0,0 +1,78 @@
+
+	if state.Family != nil {
+		res.Afi = state.Family.Afi.String()
+		res.Safi = state.Family.Safi.String()
```
This "leaks" the underlying implementation details to API response. Other implementations may have a different AFI/SAFI string representation. I think we want to have an implementation-agnostic string AFI/SAFI representation and return it, or we can just return AFI/SAFI ID (IANA standard number) from API and stringify it on the CLI side.

YutaroHayakawa commented on pkg/bgpv1/gobgp/state.go:
```diff
@@ -0,0 +1,78 @@
+		res.Safi = state.Family.Safi.String()
+	}
+
+	res.Enabled = state.Enabled
```
Is there any situation that it is disabled? Otherwise, we don't have to expose it I think.

YutaroHayakawa commented on cilium/cmd/bgp_peer_get.go:
```diff
@@ -0,0 +1,83 @@
+			fmt.Fprintf(w, "%d\t", afisafi.Received)
+			fmt.Fprintf(w, "%d\t", afisafi.Advertised)
+			fmt.Fprintf(w, "%s\t", strconv.FormatBool(afisafi.Enabled))
+			fmt.Fprintf(w, "\n")
```
What do you think about changing the output like this? Local AS, Peer AS, Peer Address   Session, and Uptime are always the same for the same peer, right?

```
Local AS   Peer AS   Peer Address   Session       Uptime   Family                 Received   Advertised   Enabled
65010      65010     10.0.0.1       ESTABLISHED   11m19s   AFI_IP_SAFI_UNICAST    11         1            true
                                                           AFI_IP6_SAFI_UNICAST   0          0            true
```

YutaroHayakawa commented on pkg/bgpv1/gobgp/state.go:
```diff
@@ -0,0 +1,78 @@
+		}
+
+		if peer.State != nil {
+			peerState.SessionState = peer.State.SessionState.String()
```
This "leaks" the underlying implementation details to API response. Other implementations may have a different state string representation. I think we want to have an implementation-agnostic state string representation.

YutaroHayakawa commented on pkg/bgpv1/gobgp/state.go:
```diff
@@ -0,0 +1,78 @@
+
+		if peer.State != nil {
+			peerState.SessionState = peer.State.SessionState.String()
+			peerState.AdminState = peer.State.AdminState.String()
```
We don't have any feature to set the admin state right now, so IMO, we don't have to expose this now.

YutaroHayakawa commented on api/v1/openapi.yaml:
```diff
@@ -3447,3 +3463,58 @@ definitions:
+      admin-state:
+        description: BGP peer configured administrative state
+        type: string
+      uptime:
```
I'd prefer to return an integer (with seconds let's say) instead of a formatted string from API. Also, I think it would be nice to have an explicit unit like `uptime-seconds`.

harsimran-pabla replied:
Done, returning uptime in integer. In CLI, parsing it with time.Duration, so it can be human readable.

YutaroHayakawa replied:
Great! Can you put the unit to the field name as well? Like `uptime-nanoseconds`.

harsimran-pabla replied:
Updated field name to uptime-nanoseconds

dylandreimerink reviewed (approved):
LGTM! Thanks for this work!

YutaroHayakawa commented on cilium/cmd/bgp_peer_get.go:
```diff
@@ -0,0 +1,87 @@
+				fmt.Fprintf(w, strings.Repeat("\t", 5))
+			}
+			// AFI and SAFI are concatenated for brevity
+			fmt.Fprintf(w, "%s_%s\t", afisafi.Afi, afisafi.Safi)
```
Super Nit: I'd prefer ":" or "/" as an AFI/SAFI delimiter since some SAFI consists of multiple words, and we may want to use underscore as a word delimiter (e.g., MPLS_VPN https://pkg.go.dev/github.com/osrg/gobgp/v3@v3.13.0/api#Family_Safi).

harsimran-pabla replied:
Updated

harsimran-pabla committed:
BGPv1: API changes to support peering state

Introduces new BGP state model in API, BGP Peers and BGP Families.
BGP Peers express session information and BGP Families expresses number
of AFI/SAFI routes exchanged.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
BGPv1: Expose peering state from gobgp

Exposes BGP Peering state from gobgp using API ListPeers. This change
also introduces state.go SHIM layer in bgp agent to which underlying bgp
implementation translates various BGP parameters.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
BGPv1: Expose peering state via agent cli

Introduces new bgp cli 'cilium bgp peers', which returns status of bgp peers.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

YutaroHayakawa commented on pkg/bgpv1/agent/state.go:
```diff
@@ -0,0 +1,130 @@
+	case SessionOpenSent:
+		return "open sent"
+	case SessionOpenConfirm:
+		return "open confirm"
```
Nit: Let's use underscore as a word delimiter for the consistency with address family.

qmonnet commented on api/v1/openapi.yaml:
```diff
@@ -3447,3 +3463,52 @@ definitions:
+        type: integer
+      advertised:
+        description: Number of routes advertised of this address family to the peer
+        type: integer
```
Nit: No line break at the end of file

qmonnet commented on cilium/cmd/bgp.go:
```diff
@@ -0,0 +1,16 @@
+// bgpCmd represents the bgp command
+var bgpCmd = &cobra.Command{
+	Use:   "bgp",
+	Short: "Access to bgp control plane",
```
Nit: “BGP”, uppercase?

```suggestion
	Short: "Access to BGP control plane",
```

qmonnet reviewed (changes requested):
Thanks! Looks good overall, but could you please add a commit description to each commit to motivate the changes? The commit titles are good, but I feel like we could do with more context in the git history.

Two minor, non-blocking nits below.

qmonnet reviewed (approved):
Thanks a lot!

ldelossa commented on pkg/bgpv1/gobgp/state.go:
```diff
@@ -0,0 +1,158 @@
+
+			// Uptime is time since session got established.
+			// It is calculated by difference in time from uptime timestamp till now.
+			// Time is rounded to second precision.
```
Why do we round to second precision here? 
My assumption is because we want the CLI output to present seconds?
If that is the case, this makes an assumption of how `GetPeerState` will be used, and gives me a slight code smell.

harsimran-pabla replied:
Agreed, I will leave it in nano seconds. Just for CLI output it can be rounded to second precision. In json output, it will be actual nano-second precision.

harsimran-pabla replied:
Most of the vendors give second level precision for uptime of BGP state. 

I was looking at bgp yang model - precision in the RFC is also upto seconds. 
https://www.ietf.org/archive/id/draft-ietf-idr-bgp-model-16.html 
Value to search : last-established

ldelossa reviewed (approved):
Overall, looks good. 

Made an inline comment about a possible assumption of `GetPeerState`.

nathanjsweet reviewed (approved):
LGTM for API and CLI

YutaroHayakawa reviewed (approved):
LGTM! Nice work!
